aeron-driver/src/main/c/aeron_ipc_publication.c

```diff
@@ -286,50 +288,58 @@ int aeron_ipc_publication_update_pub_pos_and_lmt(aeron_ipc_publication_t *public
                 }
             }
 
-            int64_t proposed_limit = min_sub_pos + publication->term_window_length;
-            if (proposed_limit > publication->conductor_fields.trip_limit)
+            if (max_sub_pos >  publication->conductor_fields.consumer_position)
             {
-                aeron_ipc_publication_clean_buffer(publication, min_sub_pos);
-                aeron_counter_set_release(publication->pub_lmt_position.value_addr, proposed_limit);
-                publication->conductor_fields.trip_limit = proposed_limit + publication->trip_gain;
-                work_count++;
+                publication->conductor_fields.consumer_position = max_sub_pos;
             }
 
-            publication->conductor_fields.consumer_position = max_sub_pos;
+            work_count += aeron_ipc_publication_clean_buffer(publication, min_sub_pos);
+
+            int64_t proposed_limit = min_sub_pos + publication->term_window_length;
+            if (proposed_limit >= publication->conductor_fields.trip_limit)
+            {
+                const int64_t max_allowed_gap = (int64_t)publication->mapped_raw_log.term_length << 1;
+                const int64_t term_length_mask = (int64_t)publication->mapped_raw_log.term_length - 1;
+                const int64_t new_term_based_limit_position = proposed_limit - (proposed_limit & term_length_mask);
+                const int64_t clean_position = publication->conductor_fields.clean_position;
+                const int64_t wrap_around_gap = new_term_based_limit_position - clean_position;
+                if (wrap_around_gap <= max_allowed_gap &&
+                    (wrap_around_gap < max_allowed_gap || 0 != (clean_position & term_length_mask)))
```

Comment:
I'm a little unsure of this logic.  By my reading, if `wrap_around_gap < max_allowed_gap` is true then the whole evaluation of this logic should be true, because if `x < y` then `x <= y`.  Therefore we could rewrite this as:
```
if (wrap_around_gap < max_allowed_gap ||
    (wrap_around_gap == max_allowed_gap && 0 != (clean_position & term_length_mask)))
```
However, for `wrap_around_gap == max_allowed_gap` then `clean_position` must be `new_term_based_limit_position - term_length / 2` in which case `0 != (clean_position & term_length_mask)` will always be true.  This suggests that the logic only needs to be `wrap_around_gap <= max_allowed_gap`.

I'm not sure if that is what you are going for here or if there is some additional case that you are trying to capture.

Reply:
I've re-written this logic. It now compares both values at term-boundary and handles offset separately.